File: internal/workbook/sort.go
package workbook

import (
	"bytes"
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"

	"golang.org/x/text/cases"
	"golang.org/x/text/language"

	"kanpic/internal/formula"
	"kanpic/pkg/cellrange"
)

type SortKey struct {
	Column    int    `json:"column"`
	Direction string `json:"direction"`
}

type SortOptions struct {
	Keys          []SortKey `json:"keys"`
	HeaderRows    int       `json:"header_rows"`
	CaseSensitive bool      `json:"case_sensitive"`
	// LiteralOrder compares text character by character, the way Excel does.
	// The default reads the digits inside a value as a number, so 2월 comes
	// before 10월 and 항목2 before 항목10.
	LiteralOrder bool `json:"literal_order"`
}

type sortScalar struct {
	rank   int
	blank  bool
	number float64
	text   string
	truth  bool
}

type sortRow struct {
	originalRow int
	values      []sortScalar
}

// BuildSortCells deterministically materializes a stable row sort. Every cell
// in the data portion is included so old coordinates are cleared atomically,
// while formulas and ordinary styles move with their source row.
func BuildSortCells(existing []Cell, selected cellrange.Range, options SortOptions) ([]CellInput, error) {
	rows := selected.End.Row - selected.Start.Row + 1
	columns := selected.End.Column - selected.Start.Column + 1
	if rows < 2 || columns < 1 || options.HeaderRows < 0 || options.HeaderRows >= rows {
		return nil, fmt.Errorf("%w: sort range must contain at least two data rows", ErrInvalid)
	}
	dataRows := rows - options.HeaderRows
	if dataRows < 2 || dataRows > MaxSortCells || columns > MaxSortCells || dataRows > MaxSortCells/columns {
		return nil, fmt.Errorf("%w: sorted data range must contain 2 to %d cells", ErrInvalid, MaxSortCells)
	}
	if len(options.Keys) == 0 || len(options.Keys) > columns {
		return nil, fmt.Errorf("%w: sort keys must contain 1 to %d columns", ErrInvalid, columns)
	}
	directions := make([]int, len(options.Keys))
	seenColumns := make(map[int]bool, len(options.Keys))
	for index, key := range options.Keys {
		if key.Column < selected.Start.Column || key.Column > selected.End.Column || seenColumns[key.Column] {
			return nil, fmt.Errorf("%w: sort key column must be unique and inside the selected range", ErrInvalid)
		}
		seenColumns[key.Column] = true
		switch strings.ToLower(strings.TrimSpace(key.Direction)) {
		case "asc":
			directions[index] = 1
		case "desc":
			directions[index] = -1
		default:
			return nil, fmt.Errorf("%w: sort direction must be asc or desc", ErrInvalid)
		}
	}
	byCoordinate := make(map[string]Cell, len(existing))
	dataStart := selected.Start.Row + options.HeaderRows
	for _, cell := range existing {
		byCoordinate[coordinateKey(cell.Row, cell.Column)] = cell
		if cell.Row >= dataStart && cell.Row <= selected.End.Row && cell.Column >= selected.Start.Column && cell.Column <= selected.End.Column {
			if cell.SpillSource != "" {
				return nil, fmt.Errorf("%w: array result cell %s from %s cannot be sorted", ErrInvalid, cellrange.Address(cell.Row, cell.Column), cell.SpillSource)
			}
			if _, merged, err := CellMerge(cell); err != nil {
				return nil, err
			} else if merged {
				return nil, fmt.Errorf("%w: merged cells must be unmerged before sorting", ErrInvalid)
			}
		}
	}
	// Caser 는 여러 고루틴이 같이 쓸 수 없으므로 정렬 한 번마다 하나 만든다.
	folder := newCaseFolder()
	records := make([]sortRow, 0, dataRows)
	for row := dataStart; row <= selected.End.Row; row++ {
		record := sortRow{originalRow: row, values: make([]sortScalar, len(options.Keys))}
		for index, key := range options.Keys {
			value, err := sortableValue(byCoordinate[coordinateKey(row, key.Column)], options.CaseSensitive, folder)
			if err != nil {
				return nil, err
			}
			record.values[index] = value
		}
		records = append(records, record)
	}
	sort.SliceStable(records, func(left, right int) bool {
		for index := range options.Keys {
			comparison := compareSortScalars(records[left].values[index], records[right].values[index], options.LiteralOrder)
			if comparison != 0 {
				if records[left].values[index].blank || records[right].values[index].blank {
					return comparison < 0
				}
				return comparison*directions[index] < 0
			}
		}
		return false
	})
	inputs := make([]CellInput, 0, dataRows*columns)
	for offset, record := range records {
		destinationRow := dataStart + offset
		for column := selected.Start.Column; column <= selected.End.Column; column++ {
			source := byCoordinate[coordinateKey(record.originalRow, column)]
			formulaText := source.Formula
			if formulaText != "" {
				formulaText = formula.ShiftReferences(formulaText, destinationRow-record.originalRow, 0)
			}
			inputs = append(inputs, CellInput{Row: destinationRow, Column: column, Value: cloneJSON(source.Value), Formula: formulaText, Style: cloneJSON(source.Style), SpillSource: source.SpillSource})
		}
	}
	return inputs, nil
}

// 대소문자를 무시하고 견줄 때는 브라우저와 **같은 방식** 으로 낮춰야 한다.
// strings.ToLower 는 글자 하나씩만 보는 단순 변환이라 자바스크립트의
// toLowerCase 와 두 군데에서 갈린다.
//
//	ΟΔΟΣ  → strings.ToLower 는 οδοσ, 브라우저는 οδος (낱말 끝 시그마)
//	İ      → strings.ToLower 는 i,    브라우저는 i + 점(U+0307)
//
// 정렬은 화면이 먼저 그리고 서버가 덮으므로, 갈리는 만큼 줄이 튄다.
// x/text 의 cases.Lower 는 브라우저와 같은 유니코드 규칙을 따른다.
//
// 남는 차이는 유니코드 판 차이뿐이다. 크롬이 아는 새 글자를 x/text 가 아직
// 모르면 그 글자에서만 갈리고, x/text 를 올리면 함께 사라진다.
func newCaseFolder() cases.Caser { return cases.Lower(language.Und) }

func sortableValue(cell Cell, caseSensitive bool, folder cases.Caser) (sortScalar, error) {
	if len(bytes.TrimSpace(cell.Value)) == 0 || bytes.Equal(bytes.TrimSpace(cell.Value), []byte("null")) {
		return sortScalar{rank: 4, blank: true}, nil
	}
	decoder := json.NewDecoder(bytes.NewReader(cell.Value))
	decoder.UseNumber()
	var value any
	if err := decoder.Decode(&value); err != nil {
		return sortScalar{}, fmt.Errorf("%w: cell %s has invalid JSON value", ErrInvalid, cellrange.Address(cell.Row, cell.Column))
	}
	switch typed := value.(type) {
	case json.Number:
		number, err := strconv.ParseFloat(string(typed), 64)
		if err != nil {
			return sortScalar{}, fmt.Errorf("%w: cell %s has an invalid number", ErrInvalid, cellrange.Address(cell.Row, cell.Column))
		}
		return sortScalar{rank: 0, number: number}, nil
	case string:
		if !caseSensitive {
			typed = folder.String(typed)
		}
		return sortScalar{rank: 1, text: typed}, nil
	case bool:
		return sortScalar{rank: 2, truth: typed}, nil
	case nil:
		return sortScalar{rank: 4, blank: true}, nil
	default:
		encoded, _ := json.Marshal(typed)
		return sortScalar{rank: 3, text: string(encoded)}, nil
	}
}

func compareSortScalars(left, right sortScalar, literal bool) int {
	if left.blank != right.blank {
		if left.blank {
			return 1
		}
		return -1
	}
	if left.rank != right.rank {
		if left.rank < right.rank {
			return -1
		}
		return 1
	}
	switch left.rank {
	case 0:
		if left.number < right.number {
			return -1
		}
		if left.number > right.number {
			return 1
		}
	case 2:
		if left.truth != right.truth {
			if !left.truth {
				return -1
			}
			return 1
		}
	default:
		if literal {
			return strings.Compare(left.text, right.text)
		}
		return compareNatural(left.text, right.text)
	}
	return 0
}

File: web/src/lib/sort.ts
import { compareNatural } from './naturalOrder'
import type { Cell } from '../types'
import { cellKey } from '../state/editor'
import { shiftFormula,MAX_SORT_CELLS } from './clipboard'
import { cellMerge,type MergeRange } from './merge'

export type SortKey = { column:number; direction:'asc'|'desc' }
export type SortOptions = { keys:SortKey[]; headerRows:number; caseSensitive:boolean; literalOrder?:boolean }

type Scalar = { rank:number; blank:boolean; number?:number; text?:string; truth?:boolean }

export function materializeSort(cells:Map<string,Cell>,range:MergeRange,options:SortOptions,sheetId:string){
  const rows=range.endRow-range.startRow+1,columns=range.endColumn-range.startColumn+1,dataRows=rows-options.headerRows
  if(rows<2||columns<1||options.headerRows<0||options.headerRows>=rows||dataRows<2)throw new Error('정렬 범위에는 데이터 행이 두 개 이상 있어야 합니다.')
  if(dataRows*columns>MAX_SORT_CELLS)throw new Error(`정렬은 한 번에 최대 ${MAX_SORT_CELLS.toLocaleString()}셀(${Math.floor(MAX_SORT_CELLS/Math.max(1,columns)).toLocaleString()}행 × ${columns}열)까지 가능합니다.`)
  if(options.keys.length<1||options.keys.length>columns)throw new Error('정렬 기준 열을 하나 이상 선택하세요.')
  const used=new Set<number>()
  for(const key of options.keys){if(key.column<range.startColumn||key.column>range.endColumn||used.has(key.column))throw new Error('정렬 기준 열은 선택 범위 안에서 중복 없이 지정해야 합니다.');used.add(key.column)}
  const dataStart=range.startRow+options.headerRows
  for(let row=dataStart;row<=range.endRow;row+=1)for(let column=range.startColumn;column<=range.endColumn;column+=1){const cell=cells.get(cellKey(row,column));if(cell?.spill_source)throw new Error(`${cell.spill_source} 배열 수식의 결과 셀은 정렬할 수 없습니다.`);if(cellMerge(cell))throw new Error('병합된 셀은 병합 해제 후 정렬할 수 있습니다.')}
  const records=Array.from({length:dataRows},(_,offset)=>{const originalRow=dataStart+offset;return{originalRow,values:options.keys.map(key=>scalar(cells.get(cellKey(originalRow,key.column))?.value,options.caseSensitive))}})
  records.sort((left,right)=>{for(let index=0;index<options.keys.length;index+=1){const comparison=compare(left.values[index],right.values[index],options.literalOrder===true);if(comparison!==0){if(left.values[index].blank||right.values[index].blank)return comparison;return comparison*(options.keys[index].direction==='asc'?1:-1)}}return left.originalRow-right.originalRow})
  const updatedAt=new Date().toISOString(),result:Cell[]=[]
  records.forEach((record,offset)=>{const destinationRow=dataStart+offset;for(let column=range.startColumn;column<=range.endColumn;column+=1){const source=cells.get(cellKey(record.originalRow,column));const formula=source?.formula?shiftFormula(source.formula,destinationRow-record.originalRow,0):undefined;result.push({sheet_id:sheetId,row:destinationRow,column,value:formula?undefined:source?.value,formula,style:source?.style?{...source.style}:undefined,updated_at:updatedAt})}})
  return result
}

function scalar(value:unknown,caseSensitive:boolean):Scalar{
  if(value===undefined||value===null)return{rank:4,blank:true}
  if(typeof value==='number')return{rank:0,blank:false,number:value}
  if(typeof value==='string')return{rank:1,blank:false,text:caseSensitive?value:value.toLowerCase()}
  if(typeof value==='boolean')return{rank:2,blank:false,truth:value}
  return{rank:3,blank:false,text:JSON.stringify(value)}
}

function compare(left:Scalar,right:Scalar,literal:boolean){
  if(left.blank!==right.blank)return left.blank?1:-1
  if(left.rank!==right.rank)return left.rank-right.rank
  if(left.rank===0)return left.number!<right.number!?-1:left.number!>right.number!?1:0
  if(left.rank===2)return left.truth===right.truth?0:left.truth?1:-1
  if(literal)return left.text!<right.text!?-1:left.text!>right.text!?1:0
  return compareNatural(left.text??'',right.text??'')
}
